packages/@lwc/template-compiler/src/shared/ir.ts

```diff
@@ -90,19 +90,33 @@ export function isComponentProp(identifier: TemplateIdentifier, node?: IRNode):
     return isComponentProp(identifier, node.parent);
 }
 
-export function isBoundToIterator(identifier: TemplateIdentifier, node?: IRNode): boolean {
+const memoizedIterators = new WeakMap<IRNode, string | null>();
+
+function getIteratorName(node?: IRNode): string | null {
     if (!node) {
-        return false;
-    }
+        return null;
+    } else {
+        let iteratorName = memoizedIterators.get(node);
+        if (iteratorName !== undefined) {
+            return iteratorName;
+        }
 
-    // Make sure the identifier is not bound to any iteration variable
-    if (isElement(node)) {
-        const { forOf } = node;
-        if (forOf) {
-            return Boolean(forOf.iterator.name === identifier.name);
+        if (isElement(node)) {
+            const { forOf } = node;
+            if (forOf) {
+                memoizedIterators.set(node, forOf.iterator.name);
+
+                return forOf.iterator.name;
+            }
         }
+
+        iteratorName = getIteratorName(node.parent);
+        memoizedIterators.set(node, iteratorName);
+
+        return iteratorName;
     }
+}
 
-    // Delegate to parent component if no binding is found at this point
-    return isBoundToIterator(identifier, node.parent);
+export function isBoundToIterator(identifier: TemplateIdentifier, node?: IRNode): boolean {
```

Comment:
Why making `node` optional?

Reply:
because this function is recursive (`node.parent`) and the stop condition is `!node`. though i can convert it to iterative (i will)